### PR TITLE
Created Classes

### PR DESCRIPTION
**Pull request by XedinUnknown**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -11,13 +11,21 @@
...
         "phpunit/phpunit": "^4.8",
         "ptrofimov/xpmock": "^1.1",
         "dhii/php-cs-fixer-config": "dev-php-5.3",
-        "codeclimate/php-test-reporter": "<=0.3.2"
+        "codeclimate/php-test-reporter": "<=0.3.2",
+        "dhii/factory-interface": "dev-develop",
+        "dhii/modular-interface": "dev-task/initial-interface"
+    },
+    "suggest": {
+        "dhii/factory-interface": "For delegating instantiation to a generic factory"
```
Comment (mecha) [inline composer.json]: Nice touch! :)

File: src/AbstractFactory.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * @since [*next-version*]
+ */
+abstract class AbstractFactory
+{
+    /**
+     * Parameter-less constructor.
+     *
+     * Invoke this in actual constructor.
+     *
+     * @@since [*next-version*]
```
Comment (mecha) [inline src/AbstractFactory.php]: Double `@`

File: src/AbstractFactory.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    protected function __construct()
```
Comment (mecha) [inline src/AbstractFactory.php]: This has two underscores. Is this intentional?

**Review by mecha — DISMISSED** [review]

**Comment by mecha** [comment]
@XedinUnknown I approve of this after the latest fixes. Are you planning to add tests?

File: src/AbstractDelegatingFactory.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * Deduces the service ID to pass to the generic factory for module creation.
+     *
+     * @since [*next-version*]
+     *
+     * @return string The service ID of the module service.
+     */
+    abstract protected function _getModuleServiceId($config = array());
```
Comment (mecha) [inline src/AbstractDelegatingFactory.php]: Undocumented `$config` parameter.

File: src/AbstractDelegatingFactory.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * Creates a new module factory exception.
+     *
+     * @since [*next-version*]
+     *
+     * @return ModuleFactoryExceptionInterface The new exception.
+     */
+    abstract protected function _createModuleFactoryException($message, ModuleFactoryInterface $factory = null, Exception $inner = null);
```
Comment (mecha) [inline src/AbstractDelegatingFactory.php]: All parameters are undocumented.

File: src/AbstractDelegatingFactory.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * @since [*next-version*]
+     *
+     * @param string                      $message The error message.
+     * @param ModuleFactoryInterface|null $factory The module factory associated with the error, if any.
+     * @param Exception                   $inner   The inner exception, if any.
+     *
+     * @return CouldNotMakeModuleExceptionInterface The new exception.
+     */
+    abstract protected function _createCouldNotMakeModuleException($message, ModuleFactoryInterface $factory = null, $moduleKey = null, $moduleConfig = null, Exception $inner = null);
```
Comment (mecha) [inline src/AbstractDelegatingFactory.php]: The `$moduleKey` and `$moduleConfig` parameters are undocumented.

**Review by mecha — DISMISSED** [review]

File: src/AbstractFactory.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    /**
+     * Creates a new module instance.
+     *
+     * @since [*next-version*]
+     *
+     * @return ModuleInterface the new module
+     */
+    abstract protected function _makeModule();
```
Comment (mecha) [inline src/AbstractFactory.php]: Missing `$config` parameter. Is this intentional?

FYI this results in:

```
Fatal error: Declaration of AbstractDelegatingFactory::_makeModule() must be compatible with AbstractFactory::_makeModule()
```

**Comment by mecha** [comment]
@XedinUnknown [This one](https://github.com/RebelCodecom/module-factory-abstract/pull/1/files#r123463098) remained unaddressed.

**Review by mecha — APPROVED** [review]
LGTM

**Comment by XedinUnknown** [comment]
@mecha, merged. Please depend on `develop`.

**Pull request merged**